lib/espresso-runner.js

```diff
@@ -29,40 +29,28 @@ class EspressoRunner {
     this.serverLaunchTimeout = opts.serverLaunchTimeout || ESPRESSO_SERVER_LAUNCH_TIMEOUT;
   }
 
-  async shouldUninstallServer () {
-    if (this.forceEspressoRebuild || !await this.adb.fileExists(TARGET_PACKAGE_CONTAINER)) {
+  async isAppPackageChanged () {
+    if (!await this.adb.fileExists(TARGET_PACKAGE_CONTAINER)) {
+      logger.debug('The previous target application package is unknown');
       return true;
     }
-
-    const tmpRoot = await tempDir.openDir();
-    try {
-      const dstPath = path.resolve(tmpRoot, path.posix.basename(TARGET_PACKAGE_CONTAINER));
-      await this.adb.pull(TARGET_PACKAGE_CONTAINER, dstPath);
-      const previousAppPackage = await fs.readFile(dstPath, 'utf8');
-      logger.debug(`The previous target application package was '${previousAppPackage}'. ` +
-        `The current package is '${this.appPackage}'.`);
-      return previousAppPackage !== this.appPackage;
-    } finally {
-      await fs.rimraf(tmpRoot);
-    }
+    const previousAppPackage = (await this.adb.shell(['cat', TARGET_PACKAGE_CONTAINER])).trim();
```

Comment:
This will just return something like `/system/bin/sh: cat: /data/local/tmp/espresso.fake.apppackage: No such file or directory` if `TARGET_PACKAGE_CONTAINER` doesn't exist, so the whole thing could be a single system call (to `adb shell cat <>`) and then simply a check for the string (if logging is something we care about, otherwise just the `return previousAppPackage !== this.appPackage;` line).

Reply:
this will make adb.shell to an exception, since cat'ing non-exisitng file returns exit code 1. Would you prefer exception-controlled over conditional-controlled flow?

Reply:
No, exception handling would be bad for this.